File: golang/76.go
/*
	Given a string S and a string T, find the minimum window
	in S which will contain all the characters in T in complexity O(n).
*/
package leetcode

import "math"

func minWindow(s string, t string) string {
	// The result is s[head : head+gap]
	start, end, head, gap := 0, 0, -1, math.MaxInt32
	count := len(t)
	mp := make([]int, 256)
	for i := 0; i < count; i++ {
		mp[t[i]]++
	}

	for end < len(s) {
		mp[s[end]]--
		if mp[s[end]] >= 0 {
			count--
		}
		end++
		// count is zero means has include all char of t
		for count == 0 {
			if end-start < gap {
				gap = end - start
				head = start
			}
			mp[s[start]]++
			if mp[s[start]] > 0 {
				count++
			}
			start++
		}
	}
	if head == -1 {
		return ""
	} else {
		return s[head : head+gap]
	}
}

File: python/76.py
class Solution(object):
    def minWindow(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        lt = len(t)
        if len(t) > len(s):
            return ""
        lst = [0] * 256
        for i in t:
            lst[ord(i)] += 1

        l, r , head, gap = 0, 0, -1, len(s)+1
        while r < len(s):
            lst[ord(s[r])] -= 1
            if lst[ord(s[r])] >= 0: # 当前字符在t里，并且不超出它的次数
                lt -= 1
            r += 1

            # lt=0, 证明此时s[l:r]包含了t的所有字符
            while lt == 0:
                if r - l < gap: # 更新结果的左边界head和区间大小gap
                    head = l
                    gap = r - l
                # 更新左边界
                lst[ord(s[l])] += 1
                if lst[ord(s[l])] > 0:
                    lt += 1
                l += 1
        if head == -1:
            return ""
        return s[head:head+gap]
  